### PR TITLE
To show tab in stage View- C# with Razor

### PR DESCRIPTION
**Pull request by v-kalc**
* Ability to render page in tab stage view from Adaptive card action
* Ability to render page in tab stage view from Deeplinking

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -26,6 +26,7 @@
 |11|Integrate graph toolkit component in teams tab      | Microsoft Teams tab sample app for demonstrating graph toolkit component |                      | [View][tab-graph-toolkit#js]|
 |12|Device permissions      | Microsoft Teams tab sample app for demonstrating device permissions |                      | [View][tab-device-permissions#js]|
 |13|Show adaptive cards in Teams tab | Microsoft Teams tab sample code which demonstrates how to [show adaptive cards in Teams tab](https://docs.microsoft.com/microsoftteams/platform/tabs/how-to/build-adaptive-card-tabs) | [View][tab-adaptive-cards#csharp]               |                | [View][tab-adaptive-cards#js]|
+|13|Tab in stage view      | Microsoft Teams tab sample app for demonstrating stage view in tab | [View][tab-stage-view#csharp]|                 |
```
Comment (AbhijeetThakur-msft) [inline README.md:29]: Rephrase it as: Microsoft Teams tab sample app for demonstrating tab in stage view

File: samples/tab-stage-view/csharp/TabInStageView/README.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
+﻿# TabInStageView
+
```
Comment (AbhijeetThakur-msft) [inline samples/tab-stage-view/csharp/TabInStageView/README.md:2]: Let's add a link to document here and also provide some info about the sample here

File: samples/tab-stage-view/csharp/TabInStageView/README.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+Bot Framework v4 echo bot sample.
```
Comment (AbhijeetThakur-msft) [inline samples/tab-stage-view/csharp/TabInStageView/README.md:3]: Let's update this description

**Review by AbhijeetThakur-msft — CHANGES_REQUESTED** [review]

File: samples/tab-stage-view/csharp/TabInStageView/Pages/sampleTab.cshtml [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    <link rel="stylesheet" type="text/css" href="/styles/msteams-16.css"/>
+    <link rel="stylesheet" type="text/css" href="/styles/custom.css"/>
+    <script src="https://statics.teams.cdn.office.net/sdk/v1.10.0/js/MicrosoftTeams.min.js"
+        integrity="sha384-6oUzHUqESdbT3hNPDDZUa/OunUj5SoxuMXNek1Dwe6AmChzqc6EJhjVrJ93DY/Bv"
+        crossorigin="anonymous"></script>
+    <script>
+        microsoftTeams.initialize();
+
+        function createLink() {
+            let url = "https://teams.microsoft.com/l/stage/8749ff03-ae9e-4d33-88d8-bc9761ea7335/0?context=%7B%22contentUrl%22%3A%22https%3A%2F%2F2f8fec398a1e.ngrok.io%2Ftab%22%2C%22websiteUrl%22%3A%22https%3A%2F%2F2f8fec398a1e.ngrok.io%2Ftab%22%2C%22name%22%3A%22Contoso%22%7D";
```
Comment (v-chetsh) [inline samples/tab-stage-view/csharp/TabInStageView/Pages/sampleTab.cshtml:16]: This URL contains hardcoded values.
Also you have mentioned ngrok URL.
Did we mentioned in Documentation regarding this URL explanation.

File: samples/tab-stage-view/csharp/TabInStageView/Startup.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        }
+
+        public IConfiguration Configuration { get; }
+
+        // This method gets called by the runtime. Use this method to add services to the container.
+        public void ConfigureServices(IServiceCollection services)
+        {
+            services.AddControllers().AddNewtonsoftJson();
+
+            services.AddRazorPages();
```
Comment (v-chetsh) [inline samples/tab-stage-view/csharp/TabInStageView/Startup.cs:30]: Either add summary or remove blank line in between.

File: samples/tab-stage-view/csharp/TabInStageView/README.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+This bot has been created using [Bot Framework](https://dev.botframework.com), it shows how to create a simple bot that accepts input from the user and echoes it back.
```
Comment (v-chetsh) [inline samples/tab-stage-view/csharp/TabInStageView/README.md:5]: Wrong Summary
**"accepts input from the user and echoes it back."**

Kindly add proper summary.

File: samples/tab-stage-view/csharp/TabInStageView/Pages/home.cshtml [hunk 7]
```diff
@@ -0,0 +1,6 @@
+﻿@page
+
+<div class="text-center">
+    <h1 class="display-4">Tab in stage View</h1>
+    <p>Learn about <a href="https://docs.microsoft.com/aspnet/core">building Web apps with ASP.NET Core</a>.</p>
+</div>
```
Comment (v-chetsh) [inline samples/tab-stage-view/csharp/TabInStageView/Pages/home.cshtml:6]: Do we need this page?

File: samples/tab-stage-view/csharp/TabInStageView/Pages/Index.cshtml.cs [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+        public IndexModel(ILogger<IndexModel> logger)
+        {
+            _logger = logger;
+        }
+
+        public void OnGet()
+        {
+
+        }
```
Comment (v-chetsh) [inline samples/tab-stage-view/csharp/TabInStageView/Pages/Index.cshtml.cs:23]: What is the use of this class??
Remove if not in use.

File: samples/tab-stage-view/csharp/TabInStageView/Models/TabInfoAction.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+{
+    using Newtonsoft.Json;
+
+    /// <summary>
+    /// Tab info action model class.
+    /// </summary>
+    public class TabInfoAction
+    {
+        /// <summary>
+        /// Gets or sets type.
```
Comment (v-chetsh) [inline samples/tab-stage-view/csharp/TabInStageView/Models/TabInfoAction.cs:15]: What type???
Please add proper summary

File: samples/tab-stage-view/csharp/TabInStageView/Models/TabInfo.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+        public string WebsiteUrl { get; set; }
+
+        /// <summary>
+        /// Gets or sets name type.
+        /// </summary>
+        [JsonProperty("name")]
+        public string Name { get; set; }
+
+        /// <summary>
+        /// Gets or sets entity id type.
```
Comment (v-chetsh) [inline samples/tab-stage-view/csharp/TabInStageView/Models/TabInfo.cs:33]: What is the "type"
You have added it at all properties summary

File: samples/tab-stage-view/csharp/TabInStageView/Models/AdaptiveCardAction.cs [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    using Microsoft.Bot.Schema;
+    using Newtonsoft.Json;
+
+    /// <summary>
+    /// Adaptive card action model class.
+    /// </summary>
+    public class AdaptiveCardAction
+    {
+        /// <summary>
+        /// Gets or sets Ms Teams card action type.
```
Comment (v-chetsh) [inline samples/tab-stage-view/csharp/TabInStageView/Models/AdaptiveCardAction.cs:16]: Why did we added "type" in summary??

**Review by v-chetsh — COMMENTED** [review]
Added my comments, please fix.

File: samples/tab-stage-view/csharp/README.md [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+## Setup
+
+1. Run ngrok - point to port 3978
+
+```bash
+# ngrok http -host-header=rewrite 3978
+```
+
+2. Create a Bot Registration
+   In Azure portal, create a [Bot Framework registration resource](https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-authentication?view=azure-bot-service-4.0&tabs=csharp%2Caadv2).
```
Comment (v-chetsh) [inline samples/tab-stage-view/csharp/README.md:25]: This resource URL is generic, please add proper article section URL here:
https://docs.microsoft.com/en-us/azure/bot-service/bot-builder-authentication?view=azure-bot-service-4.0&tabs=csharp%2Caadv2#create-the-resource

File: samples/tab-stage-view/csharp/README.md [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+3. Modify the `manifest.json` in the `/AppPackage` folder and replace the `{{BOT-ID}}` with the id from step 2.
+
+4. Zip the contents of `AppPackage` folder into a `manifest.zip`, and use the `manifest.zip` to deploy in app store or add to Teams as in step 7.
+
+5. Modify the `/appsettings.json` and fill in the `{{ Bot Id }}` and `{{ Bot Password }}` with the id from step 2.
+
+6. Add `{{ Application Base Url }}`in appsetting.json with ngrok tunnel url or deployed application base url.
```
Comment (v-chetsh) [inline samples/tab-stage-view/csharp/README.md:33]: URL

**Review by v-chetsh — COMMENTED** [review]
Reviewed & provide few more comments

File: README.md (hunk 1 above)
Comment (v-kalc) [reply README.md]: Rephrased

File: samples/tab-stage-view/csharp/TabInStageView/README.md (hunk 3 above)
Comment (v-kalc) [reply samples/tab-stage-view/csharp/TabInStageView/README.md]: updated

File: README.md [hunk 14]
```diff
@@ -141,6 +142,7 @@
 [personal-tab#cs#mvc]:samples/tab-personal/mvc-csharp
 [tab-graph-toolkit#js]:samples/tab-graph-toolkit/nodejs
 [tab-device-permissions#js]:samples/tab-device-permissions/nodejs
+[tab-stage-view#csharp]:samples/tab-stage-view/csharp
```
[the remark this answers is not in the document]
Comment (v-kalc) [reply README.md]: removed

File: samples/tab-stage-view/csharp/README.md [hunk 15]
```diff
@@ -0,0 +1,104 @@
+﻿This App talks about the Teams tab which displays Adaptive card with CSharp.
```
[the remark this answers is not in the document]
Comment (v-kalc) [reply samples/tab-stage-view/csharp/README.md]: changed Accordingly

File: samples/tab-stage-view/csharp/TabInStageView/AppPackage/manifest.json [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+        "team",
+        "personal",
+        "groupchat"
+      ],
+      "isNotificationOnly": false
+    }
+  ],
+ "staticTabs": [
+  {
+    "entityId": "Stage view",
```
[the remark this answers is not in the document]
Comment (v-kalc) [reply samples/tab-stage-view/csharp/TabInStageView/AppPackage/manifest.json]: removed

File: samples/tab-stage-view/csharp/TabInStageView/Bots/ActivityBot.cs [hunk 17]
```diff
@@ -0,0 +1,114 @@
...
+        private readonly string _appId;
+        private readonly string _applicationBaseUrl;
+
+        public ActivityBot(IConfiguration configuration)
+        {
+            _appId = configuration["MicrosoftAppId"] ?? throw new NullReferenceException("MicrosoftAppId");
+            _applicationBaseUrl = configuration["ApplicationBaseUrl"] ?? throw new NullReferenceException("ApplicationBaseUrl");
+        }
+
+        protected override async Task OnMembersAddedAsync(IList<ChannelAccount> membersAdded, ITurnContext<IConversationUpdateActivity> turnContext, CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (v-kalc) [reply samples/tab-stage-view/csharp/TabInStageView/Bots/ActivityBot.cs:45]: Added summary

File: samples/tab-stage-view/csharp/TabInStageView/Pages/sampleTab.cshtml (hunk 4 above)
Comment (v-kalc) [reply samples/tab-stage-view/csharp/TabInStageView/Pages/sampleTab.cshtml]: removed hardcoded values

File: samples/tab-stage-view/csharp/TabInStageView/Pages/home.cshtml (hunk 7 above)
Comment (v-kalc) [reply samples/tab-stage-view/csharp/TabInStageView/Pages/home.cshtml:6]: yes we may to use this later

File: samples/tab-stage-view/csharp/TabInStageView/Pages/Index.cshtml.cs (hunk 8 above)
Comment (v-kalc) [reply samples/tab-stage-view/csharp/TabInStageView/Pages/Index.cshtml.cs]: removed

File: samples/tab-stage-view/csharp/README.md (hunk 12 above)
Comment (v-kalc) [reply samples/tab-stage-view/csharp/README.md]: updated

File: samples/tab-stage-view/csharp/README.md (hunk 13 above)
Comment (v-kalc) [reply samples/tab-stage-view/csharp/README.md]: Renamed

File: samples/tab-stage-view/csharp/TabInStageView/Models/TabInfoAction.cs (hunk 9 above)
Comment (v-kalc) [reply samples/tab-stage-view/csharp/TabInStageView/Models/TabInfoAction.cs]: updated

File: samples/tab-stage-view/csharp/TabInStageView/README.md (hunk 2 above)
Comment (v-kalc) [reply samples/tab-stage-view/csharp/TabInStageView/README.md:3]: Added link

**Review by v-chetsh — APPROVED** [review]
All looks good, approving the pr.

File: samples/tab-stage-view/csharp/TabInStageView/README.md [hunk 18]
```diff
@@ -0,0 +1,112 @@
+﻿This App talks about the Teams tab in stage view with CSharp.
```
Comment (Wajeed-msft) [inline samples/tab-stage-view/csharp/TabInStageView/README.md:1]: Could you please move this Read me file one level up to csharp folder? Please update the image paths accordingly.

**Review by Wajeed-msft — CHANGES_REQUESTED** [review]

File: samples/tab-stage-view/csharp/TabInStageView/AppPackage/manifest.json [hunk 19]
```diff
@@ -0,0 +1,58 @@
+﻿{
+  "$schema": "https://raw.githubusercontent.com/OfficeDev/microsoft-teams-app-schema/preview/DevPreview/MicrosoftTeams.schema.json",
```
Comment (Wajeed-msft) [inline samples/tab-stage-view/csharp/TabInStageView/AppPackage/manifest.json:2]: Could you please check if this feature needs Dev Preview manifest schema? If not, please change it to latest schema.

https://docs.microsoft.com/en-us/microsoftteams/platform/resources/schema/manifest-schema

File: samples/tab-stage-view/csharp/TabInStageView/README.md (hunk 18 above)
Comment (AbhijeetThakur-msft) [reply samples/tab-stage-view/csharp/TabInStageView/README.md]: Updated

File: samples/tab-stage-view/csharp/TabInStageView/AppPackage/manifest.json (hunk 19 above)
Comment (AbhijeetThakur-msft) [reply samples/tab-stage-view/csharp/TabInStageView/AppPackage/manifest.json]: changed to latest schema

**Review by Wajeed-msft — APPROVED** [review]

**Review by AbhijeetThakur-msft — APPROVED** [review]
Looks good now!

**Pull request merged**